[Gerald Tan] iP

Geralddtan opened the pull request:
# Duke
> "Your mind is for having ideas, not holding them." - David Allen [Click Here](https://dansilvestre.com/productivity-quotes/)

Duke frees your mind of having to remember things you need to do. It's truly a 
* text-based software
* that's **easy** to learn
* ~~FAST~~ **Super Fast to learn!**

And... It's *FREE*

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html).
2. double-click it. 
3. add your tasks.
4. let it manage your tasks for you 🥇 

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too! Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Geralddtan committed:
We now only have a very skeleton of the Duke class.

This needs to be changed to be able to utilise good software engineering principles by creating new classes

Extracting out common behaviours into a separate method & creating new classes for abstraction allows us to employ good swe principles

Let's edit Duke class & create new quote class (Additional random feature)

Duke class:
* Employed abstraction principles for print methods so that changes can be easily made
* Instantiated new Quote class which generates a Quote of the Day
* Utilise Scanner to get user input

Quote class: Abstracted out the generation of quotes to a separate class
* Utilise Arraylist and random number generator to randomly generate a quote

Geralddtan committed:
Add some characteristics to duke, our motivator!

Geralddtan committed:
Add functionality to store and display user tasks

Right now, we perform all commands through Duke's main method. This can be abstracted out into another method for simplicity.

Lets abstract this code out into a method performCommand which performs the user task! This method will call the necessary functions within the Duke class from now on.

Geralddtan committed:
Add a Task class which is able to keep track of its done status. Created toString representation for easy printing

Expand performCommand method to cater to instances of "mark" & "unmark"

Geralddtan committed:
Add functionality for 3 types of tasks - Todo, Deadline and Events

All 3 tasks inherits from the Task class which allows us to add features unique to each task while utilising working features
Utilise Regex for matching to extract out the users' intention for their input

Geralddtan committed:
TextUITesting

Remove code which generates random quote of the day in order to facilitate stable testing

Geralddtan committed:
Error Handling through creation of DukeException class

Explanations for each error provided in code

Geralddtan committed:
Add delete functionality

Make use of arraylist to delete Tasks easily

Previously, we throw individual exceptions based on the result of checkValidTask.
Lets move throwing of invalid task errors to checkValidTask in order to abstract it out from the other methods.
Lets make checking of valid task more robust by checking if provided value < 1

Geralddtan committed:
Reactivate quote of the day functionality

Geralddtan committed:
Change command parsing

Geralddtan committed:
Add save and load function utility

Change Task to an abstract class
Add tag attribute to task ("D", "E", "T")

Geralddtan committed:
Implement Deadline LocalDate Feature to understand user date input for deadlines

Geralddtan committed:
Merge branch 'branch-Level-7'

Geralddtan committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	.gitignore
#	src/main/java/Deadline.java
#	src/main/java/Duke.java

Geralddtan committed:
Level-8 edit to solve parsing of deadline item from saved data.txt file

Geralddtan committed:
Add more OOP classes - ui, filemanager, parser, tasklist, *commands

Geralddtan committed:
Add more OOP classes - as per previous commit

Geralddtan committed:
Reorganise classes into packages - command, managers, misc, task

Geralddtan committed:
Add JUnit test for Task and TaskList

Geralddtan committed:
Create JAR File and release on github

Geralddtan committed:
Add JavaDoc Comments

Geralddtan committed:
Change code to follow coding standard more closely

Geralddtan committed:
Add find functionality

Geralddtan committed:
Merge branch 'branch-A-JavaDoc'

Geralddtan committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/ByeCommand.java
#	src/main/java/duke/managers/TaskList.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java

Geralddtan committed:
Merge branch 'branch-A-Level-9'

Geralddtan committed:
Fix FindCommand import in Parser.java

Geralddtan committed:
Change runtesh.sh to account for new directory structure

Geralddtan committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

Geralddtan committed:
Add gradle support by merging add-gradle-support branch

Enable checkstyle, Junit tests in gradle
Perform checkstyle checks and corrections

Geralddtan committed:
Merge branch 'branch-A-Gradle'

jaegarpoon commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,58 @@
+    public Duke(String savedTasksPath) {
+        this.userTaskList = new TaskList();
+        this.fileManager = new FileManager(savedTasksPath);
+        ui = new Ui();
```
Some inconsistencies in the use of "this", understood from my TA that the practice for this mod would be to omit the use of "this" if not needed!

jaegarpoon commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,115 @@
+                ui.throwDukeException("Please enter a task number to delete!");
+            } catch (NumberFormatException err) {
+                // For cases like "delete1" without a space in between or "delete 3b" where the term is not a number
+                ui.throwDukeException("Please delete a task in the following format:\n"
```
Perhaps the exceptions could be abstracted out instead of using a string?

jaegarpoon commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,36 @@
+     * @param str input string to be printed
+     */
+    public void println(Object str) {
+        System.out.println(str);
```
May I ask if this function is needed?

jaegarpoon commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,26 @@
+import duke.managers.FileManager;
+import duke.managers.TaskList;
+
+public abstract class Command {
```
I like the abstraction of the command!

jaegarpoon commented on src/main/java/duke/managers/TaskList.java:
```diff
@@ -0,0 +1,181 @@
+    public void addTask(Task task, Boolean toPrint) {
+        this.userTaskList.add(task);
+        if (toPrint) {
+            if (task.getTag() == "T") {
```
Perhaps this could be abstracted out?

jaegarpoon commented on src/main/java/duke/misc/Quote.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.Arrays;
+import java.util.Random;
+
+public class Quote {
```
Really enjoy the addition of quotes for personalisation and motivation!

jaegarpoon reviewed (commented):
Overall the flow of events is good! Just require some minor fixes and abstractions for your consideration

Geralddtan committed:
Add GUI to Duke

Geralddtan committed:
Merge branch 'branch-GUI'

Geralddtan committed:
Edit GUI appearance

Geralddtan committed:
Merge branch 'branch-Level-10'

Geralddtan committed:
Add Assertions to TaskList Methods

Previous code base did not make use of assertions, making it vulnerable to code bugs.

Added some assert statements to ensure that task methods are logical and accurate.
* addTask() Method
* deleteTask() Method
* markTaskDone() Method
* printUserTasks() Methods

Geralddtan committed:
Improve Code Quality

Currently, there was little attention paid to good coding practices. This is problematic because it makes the code less readable and potentially harder to debug.

Following good coding practices will help greatly in future work as well as help others better understand the code.

Lets improve the code quality especially following the SLAP and Make the happy path prominent principles.

Geralddtan committed:
Merge pull request #1 from Geralddtan/branch-A-Assertions

Add Assertions to TaskList Methods

Geralddtan committed:
Merge branch 'master' of github.com:Geralddtan/ip

Geralddtan committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/managers/TaskList.java

Geralddtan committed:
Merge pull request #2 from Geralddtan/branch-A-CodeQuality

Improve Code Quality

Geralddtan committed:
Set Up CI

Geralddtan committed:
update gradle.yml

Remove IO Redirection test

Geralddtan committed:
Remind Feature Implementation

Geralddtan committed:
Merge branch 'branch-Remind-Feature'

Geralddtan committed:
Edit Remind Message

Geralddtan committed:
GUI Improvements & Feature for Bye Command (Display message before exiting)

Geralddtan committed:
Update Readme, saved Ui.png file, fix formatting issues

Geralddtan committed:
Set theme jekyll-theme-cayman

Geralddtan committed:
Edit READMe

Geralddtan committed:
Merge branch 'master' of github.com:Geralddtan/ip

Geralddtan committed:
Edit for Final JAR File Release
* Archive BaseName
* Duke Exception Content
